PowerMock tests are causing ClassCastException in assertJ best practices receipt

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
<!--
Whenever possible please try to replicate your issue with the latest versions of OpenRewrite.
The latest major and minor versions of OpenRewrite projects are usually listed here:
https://docs.openrewrite.org/reference/latest-versions-of-every-openrewrite-module
For patch releases check the GitHub Releases page for the respective project.

We release every few weeks, so it's possible that your issue has already been fixed.

If you want to try the most recent changes that haven't been fully released yet, you can check out our snapshot releases
https://docs.openrewrite.org/reference/snapshot-instructions
-->
I am using

- Maven/Gradle plugin v6.15.0
- rewrite-testing-frameworks v3.14.1

## How are you running OpenRewrite?
<!--
Are you using the Maven plugin, Gradle plugin, Moderne CLI, Moderne.io or something else?
Is your project a single module or a multi-module project?

Can you share your configuration so that we can rule out any configuration issues?

Is your project public? If so, can you share a link to it?
Code snippets can also be shared privately via [our public Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA).
-->
I am using the Maven plugin, and my project is a big project with multiple modules. That's why I excluded folders to speed up the proccess.
```xml
<plugin>
      <groupId>org.openrewrite.maven</groupId>
      <artifactId>rewrite-maven-plugin</artifactId>
      <version>6.15.0</version>
      <configuration>
        <exportDatatables>true</exportDatatables>
        <activeRecipes>
          <recipe>org.openrewrite.java.testing.assertj.Assertj</recipe>
        </activeRecipes>
        <exclusions>
          <exclusion>services/*</exclusion>
          <exclusion>services/**</exclusion>
          <exclusion>*/services/**</exclusion>
          <exclusion>*/log/**</exclusion>
          <exclusion>*/target/**</exclusion>
          <exclusion>*/datascience/**</exclusion>
        </exclusions>
      </configuration>
      <dependencies>
        <dependency>
          <groupId>org.openrewrite.recipe</groupId>
          <artifactId>rewrite-testing-frameworks</artifactId>
          <version>3.14.1</version>
        </dependency>
      </dependencies>
    </plugin>
```

## What is the smallest, simplest way to reproduce the problem?
<!--
Sometimes the logs indicate a recipe stumbled over a particular pattern of code.
If you can share a code snippet that reproduces the issue, that will help us fix it faster.
We also accept [pull requests that merely replicate an issue](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes), as a step up to a full fix.

A code snippet can be something simple like this, or similar for other languages:
-->

I think the setUp method is triggering the issue but I'm not sure.

```java
@Before
  public void setUp() throws Exception {
    PowerMockito.mockStatic(System.class);
    PowerMockito.when(System.currentTimeMillis()).thenReturn(CURRENT_TIME_MILLIS);

    PowerMockito.whenNew(Disruptor.class).withAnyArguments().thenReturn(this.disruptorMock);

    PowerMockito.doAnswer(
            invocation -> {
              this.eventHandler = invocation.getArgument(0);
              return Mockito.mock(EventHandlerGroup.class);
            })
        .when(this.disruptorMock)
        .handleEventsWith(ArgumentMatchers.<EventHandler<? super Event>>any());

    PowerMockito.when(this.disruptorMock.getRingBuffer()).thenReturn(this.ringBufferMock);

    PowerMockito.when(this.ringBufferMock.tryPublishEvent(Mockito.any()))
        .thenAnswer(
            invocation -> {
              EventTranslator<Event> translator = invocation.getArgument(0);
              translator.translateTo(this.eventMock, 0);
              return true;
            });

    PowerMockito.whenNew(CrossdeviceService.Client.class)
        .withAnyArguments()
        .thenReturn(this.crossDeviceServiceMock);

    PowerMockito.whenNew(TSocket.class)
        .withArguments(HOST, THRIFT_PORT)
        .thenReturn(this.tSocketMock, this.monitoringSocketMock);

    this.eventMock = Mockito.spy(new Event());

    PowerMockito.when(this.cdoEventMock.hasUserId()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getUserIdValue()).thenReturn("1");

    PowerMockito.when(this.cdoEventMock.hasDomain()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getDomain()).thenReturn(DOMAIN);

    PowerMockito.when(this.cdoEventMock.hasUrl()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getUrl()).thenReturn(URL);

    PowerMockito.when(this.cdoEventMock.hasSourceInfo()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getSourceInfo()).thenReturn(SOURCE_INFO);

    PowerMockito.when(this.cdoEventMock.hasIp()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getIp())
        .thenReturn(new IpAddress("192.168.1.10", "2001:db8:85a3:370:7334:8a2e:370:7334"));

    PowerMockito.when(this.cdoEventMock.hasConsentString()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getConsentString()).thenReturn(CONSENT_STRING);

    PowerMockito.when(this.cdoEventMock.hasSourceType()).thenReturn(true);
    PowerMockito.when(this.cdoEventMock.getSourceType()).thenReturn(SOURCE_TYPE);

    this.crossDeviceServer = new CrossDeviceServer(HOST, this.monitoringMock, SALT);
  }
```

## What did you expect to see?
No errors.

## What did you see instead?
A stacktrace



## What is the full stack trace of any errors you encountered?

I'm running the plugin with the following command:

<!-- When errors occur, please include the output of `--stacktrace` for Gradle or `--debug` for Maven. -->
```
MAVEN_OPTS="-Xmx32g -Xms11g" mvn rewrite:run -pl :cdo-forwarder -am -T24C
```

Resulting in this stacktrace.

<!-- A code snippet, or a description of the behavior you saw instead of the above expected result. -->
```
[ERROR] Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:6.15.0:run (default-cli) on project cdo-forwarder: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:6.15.0:run failed: Error while visiting common/cdo-forwarder/src/test/java/com/activeagent/crossdevice/forwarder/CrossDeviceServerTest.java: java.lang.ClassCastException: class org.openrewrite.java.tree.J$FieldAccess cannot be cast to class org.openrewrite.java.tree.J$Identifier (org.openrewrite.java.tree.J$FieldAccess and org.openrewrite.java.tree.J$Identifier are in unnamed module of loader org.codehaus.plexus.classworlds.realm.ClassRealm @2fe033d7)
[ERROR]   org.openrewrite.java.testing.mockito.PowerMockitoWhenNewToMockito$1.lambda$removeMockUsagesVisitor$1(PowerMockitoWhenNewToMockito.java:113)
[ERROR]   java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
[ERROR]   java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
[ERROR]   java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
[ERROR]   java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
[ERROR]   java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
[ERROR]   java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
[ERROR]   java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
[ERROR]   org.openrewrite.java.testing.mockito.PowerMockitoWhenNewToMockito$1.removeMockUsagesVisitor(PowerMockitoWhenNewToMockito.java:113)
[ERROR]   org.openrewrite.java.testing.mockito.PowerMockitoWhenNewToMockito$1.visitMethodDeclaration(PowerMockitoWhenNewToMockito.java:87)
[ERROR]   org.openrewrite.java.testing.mockito.PowerMockitoWhenNewToMockito$1.visitMethodDeclaration(PowerMockitoWhenNewToMockito.java:57)
[ERROR]   org.openrewrite.java.tree.J$MethodDeclaration.acceptJava(J.java:4036)
[ERROR]   org.openrewrite.java.tree.J.accept(J.java:60)
[ERROR]   org.openrewrite.TreeVisitor.visit(TreeVisitor.java:245)
[ERROR]   org.openrewrite.TreeVisitor.visitAndCast(TreeVisitor.java:311)
[ERROR]   org.openrewrite.java.JavaVisitor.visitRightPadded(JavaVisitor.java:1374)
[ERROR]   ...
[ERROR] -> [Help 1]
```

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->

--- a/src/main/java/org/openrewrite/java/testing/mockito/PowerMockitoWhenNewToMockito.java
+++ b/src/main/java/org/openrewrite/java/testing/mockito/PowerMockitoWhenNewToMockito.java
@@ -56,7 +56,8 @@ public class PowerMockitoWhenNewToMockito extends Recipe {
         Expression target = fieldAccess.getTarget();
         if (target instanceof J.FieldAccess) {
             return ((J.FieldAccess) target).getSimpleName();
-        } else if (target instanceof J.Identifier) {
+        }
+        if (target instanceof J.Identifier) {
             return ((J.Identifier) target).getSimpleName();
         }
         return fieldAccess.getSimpleName();

--- a/src/main/java/org/openrewrite/java/testing/mockito/PowerMockitoWhenNewToMockito.java
+++ b/src/main/java/org/openrewrite/java/testing/mockito/PowerMockitoWhenNewToMockito.java
@@ -52,6 +52,16 @@ public class PowerMockitoWhenNewToMockito extends Recipe {
         return "Replaces `PowerMockito.whenNew` calls with respective `Mockito.whenConstructed` calls.";
     }
 
+    private static String extractClassName(J.FieldAccess fieldAccess) {
+        Expression target = fieldAccess.getTarget();
+        if (target instanceof J.FieldAccess) {
+            return ((J.FieldAccess) target).getSimpleName();
+        } else if (target instanceof J.Identifier) {
+            return ((J.Identifier) target).getSimpleName();
+        }
+        return fieldAccess.getSimpleName();
+    }
+
     @Override
     public TreeVisitor<?, ExecutionContext> getVisitor() {
         return Preconditions.check(new UsesMethod<>(PM_WHEN_NEW), new JavaVisitor<ExecutionContext>() {
@@ -91,9 +101,10 @@ public class PowerMockitoWhenNewToMockito extends Recipe {
                     // onlyIfReferenced=false as `maybeAddImport` doesn't seem to find the type referred to in a try statement
                     // see https://github.com/openrewrite/rewrite/issues/5187
                     maybeAddImport("org.mockito.MockedConstruction", false);
+                    maybeAddImport("org.mockito.Mockito", false);
 
                     for (J.FieldAccess mockArgument: mockArguments) {
-                        String mockedClassName = ((J.Identifier) mockArgument.getTarget()).getSimpleName();
+                        String mockedClassName = extractClassName(mockArgument);
                         String variableNameForMock = generateVariableName("mock" + mockedClassName, updateCursor(ret), INCREMENT_NUMBER);
                         J.MethodDeclaration appliedTemplate = JavaTemplate.builder(String.format("try (MockedConstruction<%s> %s = Mockito.mockConstruction(%s.class)) { } ", mockedClassName, variableNameForMock, mockedClassName))
                                 .contextSensitive()
@@ -110,7 +121,7 @@ public class PowerMockitoWhenNewToMockito extends Recipe {
             }
 
             private JavaIsoVisitor<ExecutionContext> removeMockUsagesVisitor(List<J.FieldAccess> mockArguments, J.MethodDeclaration inMethod) {
-                Set<String> mockedClassNames = mockArguments.stream().map(fa -> ((J.Identifier) fa.getTarget()).getSimpleName()).collect(toSet());
+                Set<String> mockedClassNames = mockArguments.stream().map(PowerMockitoWhenNewToMockito::extractClassName).collect(toSet());
                 return new JavaIsoVisitor<ExecutionContext>() {
                     @Override
                     public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations multiVariable, ExecutionContext ctx) {
@@ -125,8 +136,7 @@ public class PowerMockitoWhenNewToMockito extends Recipe {
                                 J.MethodInvocation initializer = (J.MethodInvocation) varr.getInitializer();
                                 if (initializer.getArguments().size() == 1 && initializer.getArguments().get(0) instanceof J.FieldAccess) {
                                     J.FieldAccess classReference = (J.FieldAccess) initializer.getArguments().get(0);
-                                    String mockedClassName = ((J.Identifier) classReference.getTarget()).getSimpleName();
-                                    if (mockedClassNames.contains(mockedClassName)) {
+                                    if (mockedClassNames.contains(extractClassName(classReference))) {
                                         return false;
                                     }
                                 }

--- a/src/test/java/org/openrewrite/java/testing/mockito/ReplacePowerMockitoIntegrationTest.java
+++ b/src/test/java/org/openrewrite/java/testing/mockito/ReplacePowerMockitoIntegrationTest.java
@@ -20,6 +20,7 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 import org.openrewrite.DocumentExample;
 import org.openrewrite.InMemoryExecutionContext;
+import org.openrewrite.Issue;
 import org.openrewrite.java.JavaParser;
 import org.openrewrite.test.RecipeSpec;
 import org.openrewrite.test.RewriteTest;
@@ -705,6 +706,7 @@ class ReplacePowerMockitoIntegrationTest implements RewriteTest {
 
               import org.junit.jupiter.api.Test;
               import org.mockito.MockedConstruction;
+              import org.mockito.Mockito;
 
               import static org.junit.jupiter.api.Assertions.assertEquals;
 
@@ -785,6 +787,7 @@ class ReplacePowerMockitoIntegrationTest implements RewriteTest {
 
               import org.junit.jupiter.api.Test;
               import org.mockito.MockedConstruction;
+              import org.mockito.Mockito;
 
               import static org.junit.jupiter.api.Assertions.assertEquals;
 
@@ -885,6 +888,71 @@ class ReplacePowerMockitoIntegrationTest implements RewriteTest {
           )
         );
     }
+
+    @Issue("https://github.com/openrewrite/rewrite-testing-frameworks/issues/785")
+    @Test
+    void whenNewWithFieldAccess() {
+        //language=java
+        rewriteRun(
+          java(
+            """
+              import org.powermock.api.mockito.PowerMockito;
+              import static org.powermock.api.mockito.PowerMockito.*;
+
+              import org.junit.jupiter.api.Test;
+              import static org.junit.jupiter.api.Assertions.assertEquals;
+
+              class MyTest {
+                  static class Generator {
+                      public int getLuckyNumber() {
+                        return 436;
+                      }
+                  }
+                  @Test
+                  void testNumbers() throws Exception {
+                      Generator mock = mock(MyTest.Generator.class);
+                      PowerMockito.whenNew(MyTest.Generator.class).withAnyArguments().thenReturn(mock);
+
+                      Generator gen = new Generator();
+                      when(gen.getLuckyNumber()).thenReturn(504);
+
+                      assertEquals(504, gen.getLuckyNumber());
+                  }
+              }
+              """,
+            """
+              import static org.mockito.Mockito.when;
+              import static org.mockito.Mockito.mock;
+
+              import org.junit.jupiter.api.Test;
+              import org.mockito.MockedConstruction;
+              import org.mockito.Mockito;
+
+              import static org.junit.jupiter.api.Assertions.assertEquals;
+
+              class MyTest {
+                  static class Generator {
+                      public int getLuckyNumber() {
+                        return 436;
+                      }
+                  }
+
+                  @Test
+                  void testNumbers() throws Exception {
+                      try (MockedConstruction<Generator> mockGenerator = Mockito.mockConstruction(Generator.class)) {
+
+                          Generator gen = new Generator();
+                          when(gen.getLuckyNumber()).thenReturn(504);
+
+                          assertEquals(504, gen.getLuckyNumber());
+                      }
+                  }
+              }
+              """
+          )
+        );
+    }
+
 
     @ParameterizedTest
     @ValueSource(strings = {

